File: backend/app/core/upgrades/stat_modifiers/explosion_damage_1.py
from ..upgrade_implementation import UpgradeImplementation, UpgradeAppliers


class ExplosionDamage1(UpgradeImplementation):
    id = 'explosion_damage_1'
    name = 'Volatile Core'
    description = '+5 explosion damage'
    rarity = 'rare'
    type = 'stat_modifier'
    cost = 10

    def can_apply(self, state: dict) -> bool:
        return True

    def apply(self, appliers: UpgradeAppliers) -> None:
        value = 5
        appliers.modifiers.add_multiplier('bullet', 'explosionDamage', value)

    def validate(self, recorded_stats: dict) -> bool:
        return True

File: frontend/src/game/upgrades/stat_modifiers/explosion_damage_1.ts
import { Upgrade, type UpgradeDef, type ExplosionSpec } from '../Upgrade'
import { UpgradeTargetID, UpgradeStatID, RarityID, UpgradeTypeID } from '../../data/ID'

export const ExplosionDamage1Def: UpgradeDef = {
  id: "explosion_damage_1",
  name: "Volatile Core",
  description: "+5 explosion damage",
  rarity: RarityID.Rare,
  upgradeType: UpgradeTypeID.StatModifier,
  cost: 10,
  targetClass: UpgradeTargetID.Bullet,
  fieldInTargetClass: UpgradeStatID.ExplosionDamage,
  value: 5,
  isMultiplier: false,
  specificAttackType: "bullet",
  stackable: true,
  maxStacks: 99999,
  dependentOn: [{ ids: ["explosive_bullets", "explosion_on_kill"] }],
}

export class ExplosionDamage1 extends Upgrade {
  // Stat fields in the def are UI/backend metadata only — the behavior is the
  // hook below, so the default modifier-channel application is skipped.
  onApply(): void {}

  modifyExplosion(explosion: ExplosionSpec): void {
    explosion.damage += this.def.value!
  }
}
